### PR TITLE
Add support for aurora auto-discovery service for DBM integrations

### PR DESCRIPTION
**Pull request by jmeunier28**
### What does this PR do?

Adds a new listener service, `DBMAuroraListener`, which can be configured by customers who wish to automate endpoint discovery in their aurora clusters & configure these with the Database Monitoring product. The listener configuration in the `datadog.yaml` file is purposely made very generic. E.g.:

```yaml
database_monitoring:
  autodiscovery:
    aurora:
      enabled: true
```

With the listener enabled, it will poll their account for a set of aurora cluster ids (based on tag filters), and create either `postgres` or `mysql`  instances from each by resolving supported template variables. An example `auto_conf.yaml` looks like this:

```yaml
ad_identifiers:
  - _dbm_aws_aurora
init_config:
instances:
  - host: "%%host%%"
    port: "%%port%%"
    aws:
      instance_endpoint: "%%host%%"
      region: "%%extra_region%%"
      managed_authentication:
        enabled: "%%extra_managed_authentication_enabled%%"
    tags:
      - "dbclusteridentifier:%%extra_dbclusteridentifier%%"
      - "region:%%extra_region%%"
```

The feature requires the customer to create a special IAM role, with `AmazonRDSReadOnlyAccess` and associate it with the ec2 instance where the agent is being run. The agent will then poll for endpoints across all specified clusters, and detect updates as the clusters scale up / down. 

As new endpoints are discovered, they each creates a new instance of the `DBMAuroraService`. Each unique service is used to resolve the template variables for an individual check & these are all eventually reconciled in the `ConfigProvider`. This means that the customer has to define a single "instance" in their `postgres.d/conf.yaml`, and they can use the provided template variables to resolve anything that is discovered via auto-discovery. For example, 

```yaml
ad_identifiers:
  - _dbm_aws_aurora
init_config:
instances:
  - host: "%%host%%"
    port: "%%port%%"
    username: datadog
    password: ENC["/path/to/db-password"]
    dbm: true
```

If the `Listener` finds the following hosts, `host-A` `host-B`, then that would resolve to these instances:

```yaml
instances:
  - host: "host-A"
    port: 5432
    username: datadog
    password: ENC["/path/to/db-password"]
    dbm: true
 - host: "host-B"
    port: 5432
    username: datadog
    password: ENC["/path/to/db-password"]
    dbm: true
```

### Motivation

This PR is work based on the solution described in [this RFC](https://docs.google.com/document/d/18n2SlbzT0n1-cB-CUCfp_-ufStcoPx64R4aVz6TRnDQ/edit). 

Before this change, the only way to configure their [postgres](https://github.com/DataDog/integrations-core/tree/master/postgres) integrations was by defining a configuration per database host in the `postgres.d/conf.yaml` file in the agent. For example, for hosts `host-foo` and `host-bar`, the customer might define a configuration like this:

```yaml
init_config:
instances:
  - host: "host-foo"
    port: 5432
    username: "datadog"
    dbm: true
  - host: "host-bar"
    port: 5432
    username: "datadog"
    dbm: true
   ```

For every single database instance for which the customer wants to monitor, they have to list the endpoint in the `instances` block. This doesn't work well for more dynamic workloads. For example, during cloud-managed autoscaling events, which can occur during incidents, the agent would need to be updated and redeployed with the new replica instance endpoints. Since this is typically not a priority during incidents, customers are forced to fly blind or interrupt their investigation to redeploy the agent. 

Therefore, adding support for auto-discovering instance endpoints as they scale / up down allows customers to better automate their use of the database-monitoring product. This should help to drive customer engagement and make for easier onboarding. 

### Describe how to test/QA your changes

This change was tested by building the agent locally via `invoke agent.build  --python-runtimes 3` && running it on an ec2 instance. The ec2 instance was configured with the proper permissions to assume a role, that can run `describe` API calls on the desired AWS account. I configured the agent to listen for updates to an existing Aurora cluster.

I added a test `postgres.d/conf.yaml` file with the `_dbm_aws_aurora` AD Identifier. E.g.:

```yaml
ad_identifiers:
  - _dbm_aws_aurora
init_config:
instances:
  - host: "%%host%%"
    port: "%%port%%"
    username: "datadog"
    dbm: true
    aws:
      instance_endpoint: "%%host%%"
      region: "%%extra_region%%"
      managed_authentication:
        enabled: "%%extra_managed_authentication_enabled%%"
    tags:
      - "dbclusteridentifier:%%extra_dbclusteridentifier%%"
      - "region:%%extra_region%%"
```

This resulted in the expected number of aurora instances being discovered && scheduled. From the output of `agent configcheck`, you can see the integration configs are created correclty:

```
=== postgres check ===
Configuration provider: file
Configuration source: file:/etc/datadog-agent/conf.d/postgres.d/conf.yaml
Config for instance ID: postgres:15fe179629065625
aws:
  instance_endpoint: zhengda-lu-dbm-ec2-aurora-aurorapostgresdbwriter46-rycu6pmfvjfq.c7ug0vvtkhqv.us-east-1.rds.amazonaws.com
  managed_authentication:
    enabled: true
  region: us-east-1
  tags:
  - dbclusteridentifier:zhengda-lu-dbm-ec2-aurora-aurorapostgresdbbc385f18-nj1vpzlipqro
  - region:us-east-1
  - env:jmeunier-test
dbm: true
host: zhengda-lu-dbm-ec2-aurora-aurorapostgresdbwriter46-rycu6pmfvjfq.c7ug0vvtkhqv.us-east-1.rds.amazonaws.com
port: 5432
username: datadog
~
Auto-discovery IDs:
* database_monitoring_aurora
```

... and the output of `agent status` is OK for the postgres instances 
```
    postgres (16.1.0)
    -----------------
      Instance ID: postgres:15fe179629065625 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/postgres.d/conf.yaml
...
      Last Successful Execution Date : 2024-02-16 16:16:11 UTC (1708100171000)

      Instance ID: postgres:ab3e9da7db7e8b9e [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/postgres.d/conf.yaml
...
      Last Successful Execution Date : 2024-02-16 16:16:21 UTC (1708100181000)

      Instance ID: postgres:eee395ed78e56613 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/postgres.d/conf.yaml
...
      Last Successful Execution Date : 2024-02-16 16:16:18 UTC (1708100178000)
```

Lastly, I confirmed the instances were being correctly reported to the Database Monitoring UI. From there, I performed the following tests:

1. I forced a scale UP event in the cluster to determine that new services were being detected correctly 
2. I forced a scale DOWN event in the cluster to determine that services were being deleted correctly, once instances were terminated
3. Added a second cluster ID to monitor & confirmed that new instances were being created properly 

Still `TODO` would be to test this internally with the cluster agent and with K8s service annotations to confirm everything is working as expected. 

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/autodiscovery/listeners/dbm_aurora.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+			auroraCluster, err := l.awsClients[cluster.Region].GetAuroraClusterEndpoints(ids)
+			if err != nil {
+				_ = log.Error(err)
+				continue
+			}
+			discoveredServices := make(map[string]struct{})
+			for id, c := range auroraCluster {
+				for _, instance := range c.Instances {
+					if instance == nil {
+						continue
```
Comment (lu-zhengda) [inline pkg/autodiscovery/listeners/dbm_aurora.go:119]: let's add a warn log here

File: pkg/databasemonitoring/aws/aurora.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+					Values: idVals,
+				},
+			},
+		})
+	if err != nil {
+		return nil, fmt.Errorf("error describing aurora DB clusters: %v", err)
+	}
+	clusters := make(map[string]*AuroraCluster, 0)
+	for _, db := range clusterInstances.DBInstances {
+		if db.Endpoint != nil && db.DBClusterIdentifier != nil {
```
Comment (lu-zhengda) [inline pkg/databasemonitoring/aws/aurora.go:62]: once we get the db instances, i think it's safer to do one additional filtering on `DBInstanceStatus` to only get available instances.
https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/accessing-monitoring.html#Overview.DBInstance.Status

File: pkg/autodiscovery/listeners/dbm_aurora.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+// DBMAuroraListener implements database-monitoring aurora discovery
+type DBMAuroraListener struct {
+	sync.RWMutex
+	newService       chan<- Service
+	delService       chan<- Service
+	stop             chan bool
+	services         map[string]Service
+	config           dbmconfig.AutodiscoverClustersConfig
+	awsClients       map[string]aws.RDSClient
```
Comment (lu-zhengda) [inline pkg/autodiscovery/listeners/dbm_aurora.go:35]: **nitpick:** ‏`awsRdsClients`

File: pkg/databasemonitoring/aws/aurora.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+			if db.IAMDatabaseAuthenticationEnabled != nil {
+				instance.IamEnabled = *db.IAMDatabaseAuthenticationEnabled
+			}
+			// Set the port, if it is known
+			if db.Endpoint.Port != nil {
+				instance.Port = *db.Endpoint.Port
+			}
+			// Set the region, if it is known
+			if db.AvailabilityZone != nil {
+				region, err := parseAWSRegion(*db.AvailabilityZone)
```
Comment (lu-zhengda) [inline pkg/databasemonitoring/aws/aurora.go:80]: why do we need to parse the region from availability zone when region is required in the config?

Comment (jmeunier28) [reply pkg/databasemonitoring/aws/aurora.go:80]: yeah we don't need to return this here, I think I added this before I updated the config to require the `region` parameter, fixing ...

File: pkg/databasemonitoring/aws/aurora.go (hunk 2 above)
Comment (jmeunier28) [reply pkg/databasemonitoring/aws/aurora.go:62]: makes sense! should we only return `Available` or are there other statues on there where the db instance could still receive requests? Some of these `Configuring-` statues don't make it clear if the database could still take traffic or not.

Comment (lu-zhengda) [reply pkg/databasemonitoring/aws/aurora.go:62]: Good question. ideally we could return instances that are "connectable" but hard to tell from those status. I think it's safer to only return available instances to start with

Comment (jmeunier28) [reply pkg/databasemonitoring/aws/aurora.go:55]: kk, updated to do that

**Review by lu-zhengda — APPROVED** [review]

**Comment by jmeunier28** [comment]
> datadog-dogstatsd suse rpm size increase is too large:
  New package size is 80.19MB
  Stable package (7.50.1) size is 62.64MB
  Diff is 17.55MB > 10.00MB (max allowed diff)
  
  😬

File: LICENSE-3rdparty.csv [hunk 5]
```diff
@@ -485,6 +485,49 @@ core,github.com/aws/aws-sdk-go-v2/service/ssooidc/types,Apache-2.0,"Copyright 20
 core,github.com/aws/aws-sdk-go-v2/service/sts,Apache-2.0,"Copyright 2014-2015 Stripe, Inc. | Copyright 2015 Amazon.com, Inc. or its affiliates. All Rights Reserved."
 core,github.com/aws/aws-sdk-go-v2/service/sts/internal/endpoints,Apache-2.0,"Copyright 2014-2015 Stripe, Inc. | Copyright 2015 Amazon.com, Inc. or its affiliates. All Rights Reserved."
 core,github.com/aws/aws-sdk-go-v2/service/sts/types,Apache-2.0,"Copyright 2014-2015 Stripe, Inc. | Copyright 2015 Amazon.com, Inc. or its affiliates. All Rights Reserved."
+core,github.com/aws/aws-sdk-go/aws,Apache-2.0,"Copyright 2014-2015 Stripe, Inc. | Copyright 2015 Amazon.com, Inc. or its affiliates. All Rights Reserved."
```
Comment (vickenty) [inline LICENSE-3rdparty.csv:488]: It seems we already have version 2 of the SDK. Would it be possible for this feature to use it too?

Comment (jmeunier28) [reply LICENSE-3rdparty.csv:488]: Oh sorry, I thought I saw v1 in the original deps when I looked. I'll update this PR to use version 2

**Review by git-thuerk-done — APPROVED** [review]

File: go.mod [hunk 6]
```diff
@@ -346,20 +346,20 @@ require (
...
+	github.com/aws/aws-sdk-go-v2/internal/configsources v1.3.1 // indirect
+	github.com/aws/aws-sdk-go-v2/internal/endpoints/v2 v2.6.1 // indirect
+	github.com/aws/aws-sdk-go-v2/internal/ini v1.8.0 // indirect
+	github.com/aws/aws-sdk-go-v2/service/ebs v1.23.0 // indirect
+	github.com/aws/aws-sdk-go-v2/service/ec2 v1.149.0
+	github.com/aws/aws-sdk-go-v2/service/internal/presigned-url v1.11.1 // indirect
+	github.com/aws/aws-sdk-go-v2/service/sso v1.20.0 // indirect
+	github.com/aws/aws-sdk-go-v2/service/ssooidc v1.23.0 // indirect
+	github.com/aws/aws-sdk-go-v2/service/sts v1.28.0
+	github.com/aws/smithy-go v1.20.1 // indirect
```
Comment (jmeunier28) [inline go.mod:362]: I hit the bug described in this issue https://github.com/aws/aws-sdk-go-v2/issues/2370#issuecomment-1953308268, so I updated these libraries. I can try to do this more surgically if that is preferred though

File: pkg/autodiscovery/listeners/dbm_aurora.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+type DBMAuroraListener struct {
+	sync.RWMutex
+	newService       chan<- Service
+	delService       chan<- Service
+	stop             chan bool
+	services         map[string]Service
+	config           dbmconfig.AutodiscoverClustersConfig
+	awsRdsClients    map[string]aws.RDSClient
+	previousServices map[string]struct{}
+	ticks            <-chan time.Time
```
Comment (vboulineau) [inline pkg/autodiscovery/listeners/dbm_aurora.go:37]: You should indicate `ticks` is only there for unit testing.

**Review by vboulineau — COMMENTED** [review]
Overall LGTM, see comments.
`release.json` to be reverted before merge of course.

You will also need to choose when to activate your listener. You can auto activate it based on config content in `pkg/config/autodiscovery/autodiscovery.go`

File: pkg/autodiscovery/listeners/dbm_aurora.go [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+	defer func() {
+		if l.ticker != nil {
+			l.ticker.Stop()
+		}
+	}()
+	for {
+		for _, cluster := range l.config.Clusters {
+			ids := make([]string, 0)
+			ids = append(ids, cluster.ClusterIds...)
+			if _, ok := l.awsRdsClients[cluster.Region]; !ok {
```
[the remark this answers is not in the document]
Comment (jmeunier28) [reply pkg/autodiscovery/listeners/dbm_aurora.go:103]: yeah we were talking about that, and it seems reasonable to require that the agent monitoring the DB should be run in the same region. I can make that change

**Review by richl-dd — APPROVED** [review]
Looks good for WKIT owned files

File: pkg/autodiscovery/listeners/dbm_aurora.go [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+			}
+			// TODO: should we wait a certain number of run iterations before we remove instances?
+			deletedServices := findDeletedServices(l.previousServices, discoveredServices)
+			l.deleteServices(deletedServices)
+			l.previousServices = discoveredServices
+		}
+		select {
+		case <-l.stop:
+			return
+		case <-l.ticks:
```
[the remark this answers is not in the document]
Comment (jmeunier28) [reply pkg/autodiscovery/listeners/dbm_aurora.go:136]: It looks like `Listen` shouldn't block, but I did migrate the `for() select` to a new `run` function, and then isolated the discovery code to a single method

**Comment by jmeunier28** [comment]
@vboulineau thanks for the review 🙇  I responded to all of your feedback, and now I'm going to try to implement what we discussed on Friday on a separate branch. Then we can decide which approach is best. Thanks!

File: pkg/autodiscovery/configresolver/configresolver_test.go [hunk 10]
```diff
@@ -549,6 +549,30 @@ func TestResolve(t *testing.T) {
...
 		},
+		{
+			testName: "database monitoring aurora configuration resolves",
+			svc: &dummyService{
+				ID:            "dummy",
+				ADIdentifiers: []string{"database_monitoring_aurora"},
+				Hosts: map[string]string{
+					"": "my-cluster.cluster-123456789012.us-west-2.rds.amazonaws.com",
+				},
+				Ports:       []listeners.ContainerPort{{Port: 5432, Name: fmt.Sprintf("p%d", 5432)}},
```
Comment (vickenty) [inline pkg/autodiscovery/configresolver/configresolver_test.go:560]:
```suggestion
				Ports:       []listeners.ContainerPort{{Port: 5432, Name: "p5432"}},
```

**Review by vickenty — APPROVED** [review]

**Comment by jmeunier28** [comment]
closing in favor of https://github.com/DataDog/datadog-agent/pull/23216

**Pull request closed without merging**